rllib/examples/two_trainer_workflow.py

```diff
@@ -10,128 +10,146 @@
-            metrics.counters["agent_steps_trained_DQN"] += batch.env_steps()
-            return batch
-
-        # Generate common experiences.
-        rollouts = ParallelRollouts(workers, mode="bulk_sync")
-        r1, r2 = rollouts.duplicate(n=2)
-
-        # DQN sub-flow.
-        dqn_store_op = r1.for_each(SelectExperiences(["dqn_policy"])).for_each(
-            StoreToReplayBuffer(local_buffer=local_replay_buffer)
-        )
-        dqn_replay_op = (
-            Replay(local_buffer=local_replay_buffer)
-            .for_each(add_dqn_metrics)
-            .for_each(TrainOneStep(workers, policies=["dqn_policy"]))
-            .for_each(
-                UpdateTargetNetwork(
-                    workers, target_update_freq=500, policies=["dqn_policy"]
-                )
-            )
-        )
-        dqn_train_op = Concurrently(
-            [dqn_store_op, dqn_replay_op], mode="round_robin", output_indexes=[1]
-        )
+        # Update DQN's target net every 500 train steps.
+        if (
+            self._counters["agent_steps_trained_DQN"]
+            - self._counters[LAST_TARGET_UPDATE_TS]
+            >= 500
+        ):
+            self.workers.local_worker().get_policy("dqn_policy").update_target()
+            self._counters[NUM_TARGET_UPDATES] += 1
+            self._counters[LAST_TARGET_UPDATE_TS] = self._counters[
+                "agent_steps_trained_DQN"
+            ]
 
         # PPO sub-flow.
-        ppo_train_op = (
-            r2.for_each(SelectExperiences(["ppo_policy"]))
-            .combine(ConcatBatches(min_batch_size=200, count_steps_by="env_steps"))
-            .for_each(add_ppo_metrics)
-            .for_each(StandardizeFields(["advantages"]))
-            .for_each(
-                TrainOneStep(
-                    workers,
-                    policies=["ppo_policy"],
-                    num_sgd_iter=10,
-                    sgd_minibatch_size=128,
-                )
-            )
+        ppo_train_batch = SampleBatch.concat_samples(ppo_batches)
+        self._counters["agent_steps_trained_PPO"] += ppo_train_batch.agent_steps()
+        # Standardize advantages.
+        ppo_train_batch[Postprocessing.ADVANTAGES] = standardized(
+            ppo_train_batch[Postprocessing.ADVANTAGES]
         )
-
-        # Combined training flow
-        train_op = Concurrently(
-            [ppo_train_op, dqn_train_op], mode="async", output_indexes=[1]
+        print(
+            "PPO policy learning on samples from",
+            "agent steps trained",
+            ppo_train_batch.agent_steps(),
+        )
+        ppo_train_batch = MultiAgentBatch(
+            {"ppo_policy": ppo_train_batch}, ppo_train_batch.count
         )
+        ppo_train_results = train_one_step(self, ppo_train_batch, ["ppo_policy"])
 
-        return StandardMetricsReporting(train_op, workers, config)
+        # Combine results for PPO and DQN into one results dict.
+        results = dict(ppo_train_results, **dqn_train_results)
```

Comment:
this small example is actually really cool!! 2 trainers train 2 policies for different agents in a same ma env. nice.

it doesn't feel exactly right here though, we should be combining the result dict? for example, not overwriting ppo steps with dqn steps, but summing them up?
I wonder how Concurrently does it.

Reply:
No, they are not overridden. The PPO stats and DQN stats will reside under their policy ID keys.

```
results = {
"ppo_policy": [some stats],
"dqn_policy": [some other stats]
}
```